shared/helper/adm_preset_definitions_helper.cpp

```diff
@@ -0,0 +1,608 @@
+#include "adm_preset_definitions_helper.h"
+#include <adm/common_definitions.hpp>
+#include <adm/parse.hpp>
+#include <algorithm>
+#include <string>
+#include <string_view>
+#include <sstream>
+#include <vector>
+#include <helper/container_helpers.hpp>
+#include <helper/supplementary_definitions.hpp>
+#include <helper/cartesianspeakerlayouts.h>
+
+namespace {
+	std::vector<std::string_view> splitString(const std::string& input, char delimiter, size_t maxElements) {
+		std::vector<std::string_view> result;
+		std::string_view str(input);
+		size_t start = 0;
+
+		for (size_t i = 0; i < maxElements - 1; ++i) {
+			size_t end = str.find(delimiter, start);
+			if (end == std::string_view::npos) {
+				break;
+			}
+			result.push_back(str.substr(start, end - start));
+			start = end + 1;
+		}
+
+		// Add the remaining text, even if it contains more delimiters
+		result.push_back(str.substr(start));
+
+		return result;
+	}
+
+	template <typename Prop, typename Elm>
+	bool matchingProp(const Elm& elmA, const Elm& elmB) {
+		bool hasProp = elmA.template has<Prop>();
+		if (hasProp != elmB.template has<Prop>())
+			return false;
+		if (hasProp) {
+			auto valA = elmA.template get<Prop>();
+			auto valB = elmB.template get<Prop>();
+			if (valA != valB) return false;
+		}
+		return true;
+	}
+}
+
+AdmPresetDefinitionsHelper::AdmPresetDefinitionsHelper()
+{
+}
+
+AdmPresetDefinitionsHelper::~AdmPresetDefinitionsHelper()
+{
+}
+
+std::shared_ptr<AdmPresetDefinitionsHelper> AdmPresetDefinitionsHelper::getSingleton()
+{
+	static auto instance = std::make_shared<AdmPresetDefinitionsHelper>();
+	return instance;
+}
+
+AdmPresetDefinitionsHelper::ObjectHolder AdmPresetDefinitionsHelper::addPresetDefinitionObjectTo(
+	std::shared_ptr<adm::Document> document, const std::string& name, const adm::AudioPackFormatId packFormatId)
+{
+	auto audioObject = adm::AudioObject::create(adm::AudioObjectName(name));
+	auto holder = setupPresetDefinitionObject(document, audioObject, packFormatId);
+	document->add(holder.audioObject);
+	return holder;
+}
+
+AdmPresetDefinitionsHelper::ObjectHolder AdmPresetDefinitionsHelper::setupPresetDefinitionObject(
+	std::shared_ptr<adm::Document> document, std::shared_ptr<adm::AudioObject> existingAudioObject, const adm::AudioPackFormatId packFormatId, std::optional<int> forSingleCfIdValue)
+{
+	ObjectHolder holder;
+	holder.audioObject = existingAudioObject;
+
+	auto td = packFormatId.get<adm::TypeDescriptor>().get();
+	auto pfIdVal = packFormatId.get<adm::AudioPackFormatIdValue>().get();
+	auto pfData = getPackFormatData(td, pfIdVal);
+	holder.audioPackFormat = document->lookup(packFormatId);
+
+	if (!holder.audioPackFormat) {
```

Comment:
I think 76ce3fba6b2e6f2316f845e833fd22a6c55383df resolves this. `getDocumentAudioPackFormat` gets the APF if already present. If it's not and it's a common definition, it adds the common defs and then returns the APF. Otherwise, (i.e, it's a supplementary definition) it copies the tree and returns the newly create APF.
We expect `getPackFormatData()` should never be null as we're calling through `setupPresetDefinitionObject`. If it's null then it's not a preset definition - an error is expected.